pkgs/ffigen/lib/src/code_generator/objc_methods.dart

```diff
@@ -0,0 +1,224 @@
+          return false;
+        }
+
+        return true;
+      });
+}
+
+enum ObjCMethodKind {
+  method,
+  propertyGetter,
+  propertySetter,
+}
+
+class ObjCProperty {
+  final String originalName;
+  String? dartName;
+
+  ObjCProperty(this.originalName);
+}
+
+class ObjCMethod {
+  final String? dartDoc;
+  final String originalName;
+  final ObjCProperty? property;
+  Type returnType;
+  final List<ObjCMethodParam> params;
+  final ObjCMethodKind kind;
+  final bool isClass;
+  final bool isOptional;
+  bool returnsRetained = false;
+  ObjCInternalGlobal? selObject;
+  ObjCMsgSendFunc? msgSend;
+  ObjCBlock? protocolBlock;
+
+  ObjCMethod({
+    required this.originalName,
+    this.property,
+    this.dartDoc,
+    required this.kind,
+    required this.isClass,
+    required this.isOptional,
+    required this.returnType,
+    List<ObjCMethodParam>? params_,
+  }) : params = params_ ?? [];
+
+  bool get isProperty =>
+      kind == ObjCMethodKind.propertyGetter ||
+      kind == ObjCMethodKind.propertySetter;
+  bool get isRequired => !isOptional;
+  bool get isInstance => !isClass;
+
+  void addDependencies(
+    Set<Binding> dependencies,
+    ObjCBuiltInFunctions builtInFunctions, {
+    bool needMsgSend = false,
+    bool needProtocolBlock = false,
+  }) {
+    returnType.addDependencies(dependencies);
+    for (final p in params) {
+      p.type.addDependencies(dependencies);
+    }
+    selObject ??= builtInFunctions.getSelObject(originalName)
+      ..addDependencies(dependencies);
+    if (needMsgSend) {
+      msgSend ??= builtInFunctions.getMsgSendFunc(returnType, params)
+        ..addDependencies(dependencies);
+    }
+    if (needProtocolBlock) {
+      final argTypes = [
+        // First arg of the protocol block is a void pointer that we ignore.
+        PointerType(voidType),
+        ...params.map((p) => p.type),
+      ];
+      protocolBlock = ObjCBlock(
+        returnType: returnType,
+        argTypes: argTypes,
+      )..addDependencies(dependencies);
+    }
+  }
+
+  String getDartMethodName(UniqueNamer uniqueNamer) {
+    if (property != null) {
+      // A getter and a setter are allowed to have the same name, so we can't
+      // just run the name through uniqueNamer. Instead they need to share
+      // the dartName, which is run through uniqueNamer.
+      if (property!.dartName == null) {
+        property!.dartName = uniqueNamer.makeUnique(property!.originalName);
+      }
+      return property!.dartName!;
+    }
+    // Objective C methods can look like:
+    // foo
+    // foo:
+    // foo:someArgName:
+    // So replace all ':' with '_'.
+    return uniqueNamer.makeUnique(originalName.replaceAll(":", "_"));
+  }
+
+  bool sameAs(ObjCMethod other) {
+    if (originalName != other.originalName) return false;
+    if (kind != other.kind) return false;
+    if (isClass != other.isClass) return false;
+    if (isOptional != other.isOptional) return false;
```

Comment:
Wouldn't it being optional still confict if it has the same name?

Reply:
Yep, that's what this is checking. We want to catch cases where the methods only differ by their optional flag and report them to the user as an error. If someone reports that this appears in a legitimate API I can add an exception for it (ie, assume the method is required).